# Westermo Dataset Analysis

* **Author:** Patrik Goldschmidt (igoldschmidt@fit.vut.cz)
* **Project:** Network Intrusion Datasets: A Survey, Limitations, and Recommendations
* **Date:** 2024

Data source: [https://github.com/westermo/network-traffic-dataset](https://github.com/westermo/network-traffic-dataset)

In [1]:
import pandas as pd
import numpy as np
import os

pd.set_option('display.max_columns', None)

In [2]:
ROOT_DIR = '/data/data_surv/westermo/data'

DATA_PCAPS_DIR = os.path.join(ROOT_DIR, 'extended/pcaps')
DATA_CSVS_DIR = os.path.join(ROOT_DIR, 'extended/flows')

## PCAP Analysis

Merge the 3 PCAPS - left, bottom, right into one. According to documentation, they are supposed to be captured at the same period on different devices. For the purpose of this analysis, merging them is appropritate.

In [3]:
pcap_file_merged = os.path.join(DATA_PCAPS_DIR, 'merged.pcap')
pcap_files_all = [os.path.join(DATA_PCAPS_DIR, file) for file in os.listdir(DATA_PCAPS_DIR)]
pcap_files_str = ' '.join(pcap_files_all)
pcap_files_str

'/data/data_surv/westermo/data/extended/pcaps/left.pcap /data/data_surv/westermo/data/extended/pcaps/bottom.pcap /data/data_surv/westermo/data/extended/pcaps/merged.pcap /data/data_surv/westermo/data/extended/pcaps/right.pcap'

In [4]:
# Merge the PCAP into one file for duration and number of packets analysis
!mergecap -w $pcap_file_merged $pcap_files_str

mergecap: The file "/data/data_surv/westermo/data/extended/pcaps/merged.pcap" appears to be damaged or corrupt.
(pcapng: total block lengths (first 112 and second 1093339656) don't match)


In [5]:
# Analyze the merged file via capinfos
!capinfos $pcap_file_merged

File name:           /data/data_surv/westermo/data/extended/pcaps/merged.pcap
File type:           Wireshark/... - pcapng
File encapsulation:  Ethernet
File timestamp precision:  microseconds (6)
Packet size limit:   file hdr: (not set)
Number of packets:   2,245
File size:           261 kB
Data size:           185 kB
Capture duration:    1.375091 seconds
First packet time:   2023-03-10 10:29:34.078823
Last packet time:    2023-03-10 10:29:35.453914
Data byte rate:      135 kBps
Data bit rate:       1,081 kbps
Average packet size: 82.80 bytes
Average packet rate: 1,632 packets/s
SHA256:              4eff608e75e643baea8a509d6d0daeb882fae1479d149c68077d38960ced4d15
RIPEMD160:           a4bc98c33ce39393d9aba2911623cc2430576e2a
SHA1:                f4a455271fbcef078fbf4655744c19bc69d725cc
Strict time order:   True
Capture oper-sys:    Linux 5.15.0-133-generic
Capture application: Mergecap (Wireshark) 3.6.2 (Git v3.6.2 packaged as 3.6.2-2)
Number of interfaces in file: 1
Interface #0 info:


Corresponds to the span (and duration of ~1h 30m) with some additional information above.

## CSV Analysis

In [6]:
os.listdir(DATA_CSVS_DIR)

['output_left.csv', 'output_right.csv', 'output_bottom.csv']

In [7]:
# Load all CSV files into one DataFrame (similar handling as PCAPs)
csv_files_all = [os.path.join(DATA_CSVS_DIR, file) for file in os.listdir(DATA_CSVS_DIR)]
data = pd.DataFrame()

for filepath in csv_files_all:
    data = pd.concat([data, pd.read_csv(filepath)])

In [8]:
data.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
Index: 68729 entries, 0 to 55424
Data columns (total 64 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   sAddress         68729 non-null  object 
 1   rAddress         68729 non-null  object 
 2   sMACs            68729 non-null  object 
 3   rMACs            68729 non-null  object 
 4   sIPs             57887 non-null  object 
 5   rIPs             57887 non-null  object 
 6   protocol         68729 non-null  object 
 7   startDate        68729 non-null  object 
 8   endDate          68729 non-null  object 
 9   start            68729 non-null  float64
 10  end              68729 non-null  float64
 11  startOffset      68729 non-null  float64
 12  endOffset        68729 non-null  float64
 13  duration         68729 non-null  float64
 14  sPackets         68729 non-null  int64  
 15  rPackets         68729 non-null  int64  
 16  sBytesSum        68729 non-null  int64  
 17  rBytesSum        

In [9]:
data.head()

sAddress           rAddress              sMACs              rMACs  \
0  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff   
1  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff   
2  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff   
3  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff   
4  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff  b8:27:eb:15:88:9c  ff:ff:ff:ff:ff:ff   

  sIPs rIPs protocol                   startDate                     endDate  \
0  NaN  NaN      ARP  2023-03-10 10:29:34.078825  2023-03-10 10:29:34.078829   
1  NaN  NaN      ARP  2023-03-10 10:29:35.102354  2023-03-10 10:29:35.102358   
2  NaN  NaN      ARP  2023-03-10 10:29:36.126359  2023-03-10 10:29:36.126364   
3  NaN  NaN      ARP  2023-03-10 10:29:40.222634  2023-03-10 10:29:40.222638   
4  NaN  NaN      ARP  2023-03-10 10:29:41.246264  2023-03-10 10:29:41.246268   

          start           end  startOffset  endOffset  duration  sPackets  \
0  1.678441e+09  1.678441e+09     0.000000   0.000004  0.000005         4   
1  1.678441e+09  1.678441e+09     1.023529   1.023533  0.000005         4   
2  1.678441e+09  1.678441e+09     2.047534   2.047539  0.000006         4   
3  1.678441e+09  1.678441e+09     6.143809   6.143813  0.000005         4   
4  1.678441e+09  1.678441e+09     7.167439   7.167443  0.000005         4   

   rPackets  sBytesSum  rBytesSum  sBytesMax  rBytesMax  sBytesMin  rBytesMin  \
0         0        240          0       60.0        NaN       60.0        NaN   
1         0        240          0       60.0        NaN       60.0        NaN   
2         0        240          0       60.0        NaN       60.0        NaN   
3         0        240          0       60.0        NaN       60.0        NaN   
4         0        240          0       60.0        NaN       60.0        NaN   

   sBytesAvg  rBytesAvg         sLoad  rLoad  sPayloadSum  rPayloadSum  \
0       60.0        NaN  3.799636e+08    0.0            0            0   
1       60.0        NaN  3.799636e+08    0.0            0            0   
2       60.0        NaN  3.196383e+08    0.0            0            0   
3       60.0        NaN  3.987790e+08    0.0            0            0   
4       60.0        NaN  3.799636e+08    0.0            0            0   

   sPayloadMax  rPayloadMax  sPayloadMin  rPayloadMin  sPayloadAvg  \
0          0.0          NaN          0.0          NaN          0.0   
1          0.0          NaN          0.0          NaN          0.0   
2          0.0          NaN          0.0          NaN          0.0   
3          0.0          NaN          0.0          NaN          0.0   
4          0.0          NaN          0.0          NaN          0.0   

   rPayloadAvg  sInterPacketAvg  rInterPacketAvg  sttl  rttl  sAckRate  \
0          NaN         0.000001              NaN   NaN   NaN       NaN   
1          NaN         0.000001              NaN   NaN   NaN       NaN   
2          NaN         0.000002              NaN   NaN   NaN       NaN   
3          NaN         0.000001              NaN   NaN   NaN       NaN   
4          NaN         0.000001              NaN   NaN   NaN       NaN   

   rAckRate  sUrgRate  rUrgRate  sFinRate  rFinRate  sPshRate  rPshRate  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
3       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
4       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   sSynRate  rSynRate  sRstRate  rRstRate  sWinTCP  rWinTCP  sFragmentRate  \
0       NaN       NaN       NaN       NaN      NaN      NaN            NaN   
1       NaN       NaN       NaN       NaN      NaN      NaN            NaN   
2       NaN       NaN       NaN       NaN      NaN      NaN            NaN   
3       NaN       Na

In [10]:
data.describe()

start           end   startOffset     endOffset      duration  \
count  6.872900e+04  6.872900e+04  68729.000000  68729.000000  68729.000000   
mean   1.678443e+09  1.678443e+09   2747.186999   2747.501637      0.314639   
std    1.574062e+03  1.574066e+03   1574.062202   1574.065929      0.211489   
min    1.678441e+09  1.678441e+09      0.000000      0.000002      0.000001   
25%    1.678442e+09  1.678442e+09   1394.870249   1394.955187      0.005006   
50%    1.678443e+09  1.678443e+09   2743.821532   2744.232317      0.419887   
75%    1.678445e+09  1.678445e+09   4125.209604   4125.518950      0.473531   
max    1.678446e+09  1.678446e+09   5443.588241   5443.588243      0.500000   

           sPackets      rPackets     sBytesSum     rBytesSum     sBytesMax  \
count  68729.000000  68729.000000  68729.000000  68729.000000  67298.000000   
mean      38.466106     29.529762   2673.185686   2042.684136     89.627760   
std       41.413802     27.212631   2953.707989   1791.675789     84.975692   
min        0.000000      0.000000      0.000000      0.000000     36.000000   
25%        4.000000      8.000000    380.000000    488.000000     65.000000   
50%       16.000000     30.000000   1040.000000   1856.000000     65.000000   
75%       92.000000     48.000000   6560.000000   4020.000000     97.000000   
max      132.000000    774.000000  31384.000000  34048.000000   1500.000000   

          rBytesMax     sBytesMin     rBytesMin     sBytesAvg     rBytesAvg  \
count  52681.000000  67298.000000  52681.000000  67298.000000  52681.000000   
mean     111.860443     61.976686     54.541922     70.307950     68.887781   
std      201.514537     33.951491      6.240214     34.792253     21.376537   
min       36.000000     36.000000     36.000000     36.000000     36.000000   
25%       64.000000     52.000000     52.000000     60.000000     58.000000   
50%       65.000000     52.000000     52.000000     65.000000     62.200000   
75%      109.000000     65.000000     60.000000     70.500000     73.867000   
max     1412.000000    328.000000    233.000000    627.680000    294.833000   

              sLoad         rLoad   sPayloadSum   rPayloadSum   sPayloadMax  \
count  6.872900e+04  6.872900e+04  68729.000000  68729.000000  67298.000000   
mean   9.858837e+07  1.802648e+06    671.189236    529.234268     36.918393   
std    3.046313e+08  4.379169e+07    866.928168    717.767704     86.076003   
min    0.000000e+00  0.000000e+00      0.000000      0.000000      0.000000   
25%    2.519567e+04  1.524026e+04     16.000000      0.000000     13.000000   
50%    1.272813e+05  3.197424e+04    222.000000    192.000000     13.000000   
75%    9.386262e+05  7.727598e+04   1776.000000   1296.000000     45.000000   
max    7.872000e+09  4.036668e+09  28784.000000   7376.000000   1448.000000   

        rPayloadMax   sPayloadMin   rPayloadMin   sPayloadAvg   rPayloadAvg  \
count  52681.000000  67298.000000  52681.000000  67298.000000  52681.000000   
mean      60.152161      9.545663      2.906456     17.661093     17.112158   
std      201.434117     35.983947      6.131870     36.976179     21.084973   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       12.000000      0.000000      0.000000      7.000000      6.000000   
50%       13.000000      0.000000      0.000000     13.000000     10.200000   
75%       57.000000     13.000000      8.000000     18.500000     21.867000   
max     1360.000000    300.000000    205.000000    575.680000    241.500000   

       sInterPacketAvg  rInterPacketAvg     sttl          rttl      sAckRate  \
count     62892.000000     52620.000000  50827.0  52111.000000  50827.000000   
mean          0.011273         0.013121     64.0     63.318948      0.996968   
std           0.019649         0.009113      0.0      3.208340      0.023996   
min           0.000000         0.000000     64.0     39.000000      0.571000   
25%           0.000490         0.007092     64.0     6

In [11]:
# Addresses breakdown
data['sAddress'].value_counts()

sAddress
198.18.134.11        28264
198.18.134.12        18753
b8:27:eb:15:88:9c     7975
198.18.134.1          2958
198.18.134.49         1057
0.0.0.0                924
198.18.134.47          775
198.18.134.42          631
198.18.134.3           513
198.18.134.31          430
198.18.134.43          426
198.18.134.99          415
198.18.134.41          392
198.18.134.48          389
198.18.134.6           376
00:07:7c:88:6e:83      363
00:07:7c:88:6e:80      362
00:07:7c:29:de:60      362
00:07:7c:88:6e:60      362
00:07:7c:8c:43:60      362
198.18.134.38          326
b8:27:eb:d1:b7:ef      288
198.18.134.45          251
198.18.134.46          209
198.18.134.39          194
198.18.134.40          183
00:07:7c:29:de:40      181
b8:27:eb:6d:4f:4b      112
198.18.134.5           107
00:07:7c:8c:43:63      104
00:07:7c:29:de:61       93
198.18.134.4            77
198.18.134.37           76
00:07:7c:88:6e:63       74
198.18.134.50           71
00:24:9b:6d:b8:89       69
00:07:7c:8c:43:83  

In [12]:
data['rAddress'].value_counts()

rAddress
198.18.134.15        19543
198.18.134.31        19283
198.18.134.12         9332
ff:ff:ff:ff:ff:ff     8647
198.18.134.99         5160
224.0.0.22            2341
01:80:c2:00:00:0e     1629
224.0.0.1             1277
255.255.255.255        924
b8:27:eb:3e:5d:96      293
00:24:9b:6d:b8:89      154
b8:27:eb:d1:b7:ef      112
198.18.134.6             9
224.0.0.251              6
198.18.134.45            5
198.18.134.51            5
00:07:7c:8c:43:63        4
198.18.134.48            2
00:07:7c:8c:43:83        1
00:07:7c:88:6e:83        1
00:07:7c:29:de:61        1
Name: count, dtype: int64

Interesting that they mix IP addresses and MAC addresses into a common feature field. Apparently, when protocol without IP address was used (e.g., ARP) MAC address is present, while other protocols with active IP address use it in this feature value.

In [13]:
# According to documentation, NST labeling is more precise - breakdown its labels
data['NST_M_Label'].value_counts()

NST_M_Label
Normal                     54028
BAD-MISCONF-DUPLICATION     4611
BAD-MISCONF                 3889
BAD-SSH                     2968
BAD-PORTSCAN2               2179
BAD-MITM                     708
BAD-PORTSCAN1                267
GOOD-SSH                      79
Name: count, dtype: int64

### Temporal and Data Continuity Analysis

In [14]:
pd.set_option('display.float_format', lambda x: f'{x:.24f}')

In [15]:
# What's the relation between 'startDate' and 'start' feature values?
data[['startDate', 'start', 'endDate', 'end']]

startDate                               start  \
0      2023-03-10 10:29:34.078825 1678440574.078824996948242187500000   
1      2023-03-10 10:29:35.102354 1678440575.102354049682617187500000   
2      2023-03-10 10:29:36.126359 1678440576.126358985900878906250000   
3      2023-03-10 10:29:40.222634 1678440580.222634077072143554687500   
4      2023-03-10 10:29:41.246264 1678440581.246263980865478515625000   
...                           ...                                 ...   
55420  2023-03-10 11:57:10.821072 1678445830.821072101593017578125000   
55421  2023-03-10 11:59:29.972545 1678445969.972544908523559570312500   
55422  2023-03-10 11:50:21.568717 1678445421.568717002868652343750000   
55423  2023-03-10 11:57:40.766799 1678445860.766798973083496093750000   
55424  2023-03-10 11:57:36.625931 1678445856.625931024551391601562500   

                          endDate                                 end  
0      2023-03-10 10:29:34.078829 1678440574.078829050064086914062500  
1      2023-03-10 10:29:35.102358 1678440575.102358102798461914062500  
2      2023-03-10 10:29:36.126364 1678440576.126363992691040039062500  
3      2023-03-10 10:29:40.222638 1678440580.222637891769409179687500  
4      2023-03-10 10:29:41.246268 1678440581.246268033981323242187500  
...                           ...                                 ...  
55420  2023-03-10 11:57:10.821072 1678445830.821072101593017578125000  
55421  2023-03-10 11:59:29.973469 1678445969.973469018936157226562500  
55422  2023-03-10 11:50:21.568717 1678445421.568717002868652343750000  
55423  2023-03-10 11:57:40.766800 1678445860.766799926757812500000000  
55424  2023-03-10 11:57:36.626146 1678445856.626146078109741210937500  

[68729 rows x 4 columns]

Apparently, they are the same, but one is represented in a date format, the second one in an unix timestamp format. Let's utilize the `end` to estimate the duration and continuity.

In [16]:
timestamps = pd.to_datetime(data['end'], unit='s', origin='unix')
timestamps = timestamps.sort_values(ascending=False)
timestamps

4719    2023-03-10 11:00:17.667068958
55368   2023-03-10 11:00:16.495696068
55367   2023-03-10 11:00:16.482666016
55371   2023-03-10 11:00:16.470124006
55370   2023-03-10 11:00:16.466128111
                     ...             
0       2023-03-10 09:29:34.565763950
1       2023-03-10 09:29:34.500118971
0       2023-03-10 09:29:34.078829050
0       2023-03-10 09:29:34.078828096
5       2023-03-10 09:29:34.078826904
Name: end, Length: 68729, dtype: datetime64[ns]

In [17]:
# Compute the span
capture_span = timestamps.iloc[0] - timestamps.iloc[-1]
capture_span

Timedelta('0 days 01:30:43.588242054')

In [18]:
# Span corresponds to the PCAP one. Compute its 1% for continuity estimation
capture_span * 0.01

Timedelta('0 days 00:00:54.435882420')

In [19]:
# Measure capture duration and continousness
def measure_real_capture_dur(data: pd.Series, gap_max_secs: int = 300) -> float:
    """Computes total timespan of the capture. Expects iterable containing timestamps objects sorted in a descending manner"""
    total_dur = pd.Timedelta(seconds=0)
    current_dur = pd.Timedelta(seconds=0)
    cont_durations = []
    last_tstamp = data.iloc[0]
    contiguous = True

    # Iterate through the dataframe to find out gaps
    for cur_tstamp in data:
        dur_gap = last_tstamp - cur_tstamp

        if dur_gap <= pd.Timedelta(seconds=gap_max_secs):
            total_dur += dur_gap
            current_dur += dur_gap
        else:
            cont_durations.append(current_dur)
            current_dur = pd.Timedelta(seconds=0)

            if contiguous:
                contiguous = False

        last_tstamp = cur_tstamp

    # Log the final continuous block and sort them via their length
    cont_durations.append(current_dur)
    cont_durations.sort(reverse=True)

    return total_dur, contiguous, cont_durations

In [20]:
measure_real_capture_dur(timestamps, int(capture_span.total_seconds() * 0.01))

(Timedelta('0 days 01:30:43.588242054'),
 True,
 [Timedelta('0 days 01:30:43.588242054')])

As confirmed by PCAP, this capture is continuous and lasts approximately 1.5 hours.